arduino_code/openflexure_nano_motor_controller/openflexure_nano_motor_controller.ino

```diff
@@ -137,6 +313,32 @@ void move_axes(long displacement[n_motors]){
     unsigned long start = micros();
     float final_scaled_t = (float) max_steps * min_step_delay; //NB total time taken will be final_scaled_t + 2*ramp_time
     while(!finished){
+      #if defined(ENDSTOPS_MIN) || defined(ENDSTOPS_MAX)
+        int endstop_break=0;
+        EACH_MOTOR{
+          if(endstopTriggered(i,dir[i]*(displacement[i]-distance_moved[i])))
+            endstop_break=dir[i]*(i+1);
+        }
+        if(endstop_break!=0){
+          Serial.print("Endstop hit:");
+          Serial.println(endstop_break);
+          //if we have both min/max endstops, axis_max is adjusted to the correct value
+          //if we only have min, we go from 0 -> predefined axis_max
+          //if we only have max, we go from 0 -> predefined axis_max
+          //the predefined axis_max are the travel distances in steps
+          if(endstop_break<0) //min is always 0
+            current_pos[-endstop_break-1]=0;
+          else //max
+            #if defined(ENDSTOPS_MIN) && defined(ENDSTOPS_MAX)
+              axis_max[endstop_break-1]=current_pos[endstop_break-1];
+            #else
+              current_pos[endstop_break-1]=axis_max[endstop_break-1];
+            #endif
+          ret=endstop_break;
+          goto movedone; //this is the proper use for goto
```

Comment:
While I agree this is the proper use of goto, I'll politely point out that you could achieve exactly the same thing with a break statement...